### PR TITLE
SAMA5D2x SDMMC peripheral support (SD Card driver)

### PR DESCRIPTION
**Pull request by adamfeuer**
## Summary

- SDMMC peripheral support for [SAMA5D2](http://ww1.microchip.com/downloads/en/devicedoc/ds60001476b.pdf) series microprocessors from Microchip
- provides an MMC/SD Card driver that works for reading and writing
- can be used with or without DMA
- UHS_SDR50, UHS_DDR50, and UHS_SDR104 high-speed modes are supported, with  
  read/write speeds of 400Mbit/s in UHS_DDR50 (50MHz Double Data Rate) mode.
- This is a port of the NuttX imxrt_usdhc.c driver, as the SDMMC peripheral
  uses a similar Cadence IP block.
- Basic instructions are in the README.txt.

## Impact

- None if you don't enable the SDMMC peripheral using the Kconfig system.
- Enables reading and writing to MMC or SD Cards. Only SD Cards have been tested.
- VFAT filesystem has been tested and works. On the SAMA5D2-XULT board, reading and writing to a bootable SD Card is supported.

## Testing

- Only tested on the SAMA5D2-XULT board from Microchip.
- Only manual read and write testing on a VFAT-formatted memory card has been done.
- Driver was developed with extensive use of a logic analyzer to ensure the SDMMC driver was functioning correctly.
- Only UHS_DDR50 mode was tested, as all my SD Cards support only that.

## How to Test

1. Format an SD Card with VFAT format. 
1. Put some text files on the card.
1. Insert the card into the SAMA5D2-XULT SD Card slot.
1. Build the sama5d2-xult:sdmmcnsh configuration
1. Load NuttX onto the board
1. /mnt/mmcsd1 should have your files. You can cd, ls, cp, rm, and cat
   to test the filesystem.

## TODO / Limitations

- e.MMC should work, but was not tested
- Performance tests have not been done
- UHS_SDR50 and UHS_SDR104 modes were not tested as I don't have any
  SD Cards that support those modes.

## Notes

- nxstyle shows some formatting errors, but I couldn't find good ways to make them go away without mangling the files. I welcome ideas on how to better format the code so nxstyle passes!

### CODE REVIEW

**Comment by adamfeuer** [comment]
@btashton would you be willing to help me find someone who can review this? 

I'm working to get [Groboard's GiantBoard](https://groboards.com/giant-board/) and the [ACME RoadRunner module](https://www.acmesystems.it/roadrunner) supported and added to NuttX— they're both SAMA5 boards so having SDMMC support would help. I know it's a lot of code, not sure how to avoid that.

File: boards/arm/sama5/sama5d2-xult/configs/sdmmcnsh/defconfig [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+CONFIG_SERIAL_TERMIOS=y
+CONFIG_SIG_DEFAULT=y
+CONFIG_SIG_EVTHREAD=y
+CONFIG_START_DAY=31
+CONFIG_START_MONTH=7
+CONFIG_START_YEAR=2014
+CONFIG_SYMTAB_ORDEREDBYNAME=y
+CONFIG_SYSLOG_TIMESTAMP=y
+CONFIG_SYSTEM_NSH=y
+CONFIG_SYSTEM_NSH_CXXINITIALIZE=y
```
Comment (Ouss4) [inline boards/arm/sama5/sama5d2-xult/configs/sdmmcnsh/defconfig:136]: This config doesn't exist anymore.

**Comment by adamfeuer** [comment]
@Ouss4 Re: headers, will fix that. Thank you.

File: boards/arm/sama5/sama5d2-xult/configs/sdmmcnsh/defconfig (hunk 1 above)
Comment (adamfeuer) [reply boards/arm/sama5/sama5d2-xult/configs/sdmmcnsh/defconfig:136]: @Ouss4 Ok, will remove. I'll run it through make menuconfig again after merging in master.

**Comment by adamfeuer** [comment]
@Ouss4 Ok, I changed the headers on the new files to Apache License 2.0, and removed the obsolete config value. Anything else to do here?

**Comment by adamfeuer** [comment]
@Ouss4 BTW the two main files, [sam_sdmmc.h](https://github.com/starcat-io/incubator-nuttx/blob/feature/sama5d27-sdmmc-support/arch/arm/src/sama5/hardware/sam_sdmmc.h) and [sam_sdmmc.c](https://github.com/starcat-io/incubator-nuttx/blob/feature/sama5d27-sdmmc-support/arch/arm/src/sama5/sam_sdmmc.c) are ported from [imxrt_usdhc.h](https://github.com/starcat-io/incubator-nuttx/blob/feature/sama5d27-sdmmc-support/arch/arm/src/imxrt/hardware/imxrt_usdhc.h) and [imxrt_usdhc.c](https://github.com/starcat-io/incubator-nuttx/blob/feature/sama5d27-sdmmc-support/arch/arm/src/imxrt/imxrt_usdhc.c).

File: boards/arm/sama5/sama5d2-xult/configs/sdmmcnsh/defconfig [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+CONFIG_BOARD_LOOPSPERMSEC=65775
+CONFIG_BOOT_RUNFROMSDRAM=y
+CONFIG_BUILTIN=y
+CONFIG_CONSOLE_SYSLOG=y
+CONFIG_DEBUG_FULLOPT=y
+CONFIG_DEBUG_SYMBOLS=y
+CONFIG_DEV_LOOP=y
+CONFIG_DEV_ZERO=y
+CONFIG_EXAMPLES_HELLO=y
+CONFIG_EXAMPLES_PARTITION=y
```
Comment (Ouss4) [inline boards/arm/sama5/sama5d2-xult/configs/sdmmcnsh/defconfig:38]: Doesn't exist anymore?

File: boards/arm/sama5/sama5d2-xult/configs/sdmmcnsh/defconfig [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+CONFIG_I2S=y
+CONFIG_INTELHEX_BINARY=y
+CONFIG_IOB_NBUFFERS=72
+CONFIG_IOB_THROTTLE=16
+CONFIG_MAX_TASKS=16
+CONFIG_MAX_WDOGPARMS=2
+CONFIG_NETDB_BUFSIZE=128
+CONFIG_NETINIT_NETLOCAL=y
+CONFIG_NETINIT_NOMAC=y
+CONFIG_NETUTILS_NETLIB_GENERICURLPARSER=y
```
Comment (Ouss4) [inline boards/arm/sama5/sama5d2-xult/configs/sdmmcnsh/defconfig:61]: Doesn't exist anymore?

File: boards/arm/sama5/sama5d2-xult/configs/sdmmcnsh/defconfig (hunk 2 above)
Comment (adamfeuer) [reply boards/arm/sama5/sama5d2-xult/configs/sdmmcnsh/defconfig:38]: @Ouss4 will remove

File: boards/arm/sama5/sama5d2-xult/configs/sdmmcnsh/defconfig (hunk 3 above)
Comment (adamfeuer) [reply boards/arm/sama5/sama5d2-xult/configs/sdmmcnsh/defconfig:61]: @Ouss4 Will remove

**Comment by adamfeuer** [comment]
@Ouss4 Thanks, I pushed changes to remove the obsolete config values, I'll watch the CI and try to resolve any other problems.

**Comment by adamfeuer** [comment]
@Ouss4 Ok, looks like all the non-nxstyle checks are passing. I ran nxstyle locally and fixed all the problems I could before submitting this PR; the remaining nxstyle issues I don't have good ideas about how to fix... if you have suggestions, please let me know!

**Comment by Ouss4** [comment]
>the remaining nxstyle issues I don't have good ideas about how to fix... if you have suggestions, please let me know!

For header file the "error: Invalid section for this file type" is a section named "Public Function" that needs to be "Public Function Prototypes".  There are also a couple of long line and "too many blank lines" that can easily be fixed.  For the rest, I don't see how to fix them, it's hard to make nxstyle happy with register definition files.

**Comment by adamfeuer** [comment]
@Ouss4 Oops, I am sorry I missed the easy ones. I'll fix the header name, the long lines, and the blank lines. 

For the rest of the nxstyle errors, what do you suggest? Should we just ignore them?

**Comment by Ouss4** [comment]
> For the rest of the nxstyle errors, what do you suggest? Should we just ignore them?

I don't see a way to make nxstyle happy without messing up the file, so that would be my suggestion.
One last thing, could you please squash the commits if you are happy with the changes?

**Comment by adamfeuer** [comment]
@Ouss4 Thank you. I will run one more set of manual tests on the SAMA5D2-XULT board, and also wait for the current set of CI builds to finish. If that all goes well, I'll squash and force-push, and then let you know I've done it.

**Comment by adamfeuer** [comment]
@Ouss4 I retested on hardware and force-pushed. Now it looks like the macOS (arm-12), (mips-riscv-x86-xtensa), and (sim) builds failed. But the Linux versions succeeded. Is this expected?

If these failures are unexpected— will you let me know how to debug? I don't see any detail from the CI runs.

If these failures are ok, then this PR is ready to merge.

**Comment by Ouss4** [comment]
@adamfeuer Yes, sometimes some of the macOS checks don't make it to the end.  It's Github things apparently.

**Comment by btashton** [comment]
> @adamfeuer Yes, sometimes some of the macOS checks don't make it to the end.  It's Github things apparently.

I actually re-kicked the build a while back and they are passing now. For some reason the old canceled jobs are showing up on this run. But you will also see the new Mac jobs that passed. 

All that is failing is the style check.

**Comment by adamfeuer** [comment]
@btashton Thanks for doing that. How does one restart the build? Is that something I can do without pushing a new change?

**Comment by adamfeuer** [comment]
@Ouss4 This is ready to merge now­— the only thing that's not passing is the style check for some header files, as we discussed it's hard to get nxstyle to be happy with them.

**Comment by Ouss4** [comment]
> How does one restart the build? Is that something I can do without pushing a new change?

In the checks tab, there is a button at the right that says "Re-run jobs".

**Comment by adamfeuer** [comment]
@Ouss4 You're welcome.

By the way, the "Re-run jobs" button doesn't show up for me. Maybe because I'm not a member of the NuttX Github team?

**Pull request merged**